rfd/0163-vnet.md

```diff
@@ -0,0 +1,971 @@
+Configuring a custom IP range at the cluster level would look like:
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  cidr_range: "100.64.0.0/10"
+```
+
+#### IPv6
+
+VNet will assign apps Unique Local IPv6 Unicast Addresses (ULA) as defined in
+[RFC 4193](https://datatracker.ietf.org/doc/html/rfc4193).
+They have the following format:
+
+      | 7 bits |1|  40 bits   |  16 bits  |          64 bits           |
+      +--------+-+------------+-----------+----------------------------+
+      | Prefix |L| Global ID  | Subnet ID |        Interface ID        |
+      +--------+-+------------+-----------+----------------------------+
+
+The Prefix + L will always be the ULA prefix `fd`.
+
+The Global ID will be randomly generated on VNet startup by pulling 40 bits from
+`crypto/rand`.
+
+The Subnet ID will be hardcoded to 0.
+
+The 64-bit Interface ID will be the IPv4 address assigned to the app, prefixed
+with zeros.
+
+An example IPv6 VNet address with a Global ID of `0xec146c95f6` and a related
+IPv4 address of `100.64.0.2` would look like `fdec:146c:95f6::6440:2`.
+A route would be configured on the host for the `fdec:146c:95f6::/64` CIDR
+range.
+The CIDR range will not be configurable for IPv6.
+
+#### IP access without DNS
+
+If a client wishes to access a Teleport app over VNet but their client does not
+accept a domain name and only an IP, there are two ways this can be handled.
+
+The first option is the user can first ping the domain name to cause VNet to
+assign an IP to the app, and then they may use that IP.
+
+```shell
+$ ping api.teleport.example.com
+PING api.teleport.example.com (100.64.0.4): 56 data bytes
+$ curl 100.64.0.4
+```
+
+Another option would be to support static configuration of an IP that will
+always be assigned to an app.
+This may be desirable by some users used to accessing an app at a static IP, but
+the implementation would be more challenging and fragile so we'll prefer not to
+support this in the initial implementation.
+For example, it's not clear what to do when a new app is added with a static IP
+address that has already been assigned to another app.
+VNet would also need to handle incoming TCP connections on all addresses and
+then search all clusters for an app with a static IP matching that address, or
+else we would need to maintain an up to date set of all apps in all clusters.
+
+### TCP ports
+
+Local app proxies today each listen on a unique TCP port on `127.0.0.1`.
+With VNet we will assign a unique IP address to each app with the full TCP port
+range available.
+
+The reason we need to assign an IP per app, instead of having many apps on the
+same IP but different ports, is that DNS only assigns an IP to each domain name
+with no port, so there would be no good way for the client to discover the port
+we assigned to the app.
+
+To keep configuration as simple as possible, and because we can, the default
+behavior of VNet will be to proxy TCP connections from ANY port to the app
```

Comment:
Since apps are currently configured with a port, wont this potentially expose services on the system which may not have been intended to be exposed?

We can make the port assignments 1:1, but I think we should only forward ports which are configured on the application.

Reply:
right, I didn't really make this clear, but the idea is that we still only proxy *to* the configured port for the app, but VNet locally will accept connections on any port.

E.g. if I have an app configured like:

```
kind: app
version: v3
metadata:
  name: api
spec:
  public_addr: api.teleport.example.com
  uri: http://127.0.0.1:8080
```

On the client side running VNet I can run `curl api.teleport.example.com:8080` or `curl api.teleport.example.com:80` (or any other port) and both will work and proxy to port 8080 of localhost wherever the app service is running

Reply:
What if in the future we wanted to support multiple ports to be configured per-application?  I think having the client match the port to be forwarded is pretty straight forward, and likely easier to manage in the future.

But no security concerns as long as we are not allowing arbitrary ports to be reached.

Reply:
I think it would be pretty common that you wouldn't want the local port in the URI to match the port that clients want to connect to. Like let's say you're running the app service on a server with 3 postgres instances running on ports 5001, 5002, 5003, but all clients want to connect on port 5432 to each postgres instance. Or you have an HTTP app listening locally on :8080 but you want clients to connect to :80. It's also possible to have multiple apps with the same name with different ports in the URI, I think we currently select one arbitrarily.

So I'm pretty sure we don't want to tie clients to connecting to the port from the URI. If we want to support multiple ports for the same app in the future, I think it will require the app spec to be extended, and we can handle it in VNet at that time